File: apps/android/app/src/main/java/com/termirust/mobile/replication/EnrollmentViewModel.kt
package com.termirust.mobile.replication

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.termirust.replication.security.MobileReplicationException
import com.termirust.replication.security.ReplicationStorageException
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

enum class EnrollmentProblem { LOCKED, BUSY, INVALID, CHANGED, RECOVERY, CONFIGURED, UNAVAILABLE }
data class EnrollmentState(
    val snapshot: EnrollmentSnapshot = EnrollmentSnapshot(),
    val loading: Boolean = true,
    val problem: EnrollmentProblem? = null,
    val bundleReview: EnrollmentBundleReview? = null,
)

class EnrollmentViewModel(private val repository: EnrollmentRepository) : ViewModel() {
    private val mutable = MutableStateFlow(EnrollmentState())
    val state = mutable.asStateFlow()
    private var running = false
    private var reviewGeneration = 0L

    init { reload() }
    fun reload() = operate { repository.load() }
    fun recover() {
        if (state.value.problem != EnrollmentProblem.RECOVERY || state.value.snapshot.cancellationPending) return
        operate { repository.recover() }
    }
    fun prepare() {
        if (state.value.problem != null || state.value.snapshot.request != null || state.value.snapshot.configured) return
        operate { repository.prepare() }
    }

    fun discardBundleReview() {
        reviewGeneration++
        mutable.value = mutable.value.copy(bundleReview = null)
    }

    fun reviewBundle(bytes: ByteArray) {
        val request = state.value.snapshot.request?.copyOf() ?: return
        if (running || state.value.snapshot.cancellationPending || state.value.problem != null) return
        val captured = bytes.copyOf()
        val generation = ++reviewGeneration
        running = true
        mutable.value = mutable.value.copy(loading = true, bundleReview = null)
        viewModelScope.launch {
            try {
                val review = repository.reviewBundle(request, captured)
                if (generation == reviewGeneration) mutable.value = mutable.value.copy(bundleReview = review)
            } catch (cancelled: CancellationException) { throw cancelled }
            catch (error: Exception) {
                if (generation == reviewGeneration) mutable.value = mutable.value.copy(problem = problemFor(error))
            } finally {
                running = false
                mutable.value = mutable.value.copy(loading = false)
            }
        }
    }

    fun acceptBundle(verifiedCode: String) {
        val review = state.value.bundleReview ?: return
        if (running || verifiedCode != review.code) return
        operate { repository.acceptBundle(review, verifiedCode) }
    }

    private fun problemFor(error: Exception) = when (error) {
        is MobileReplicationException.Locked, is ReplicationStorageException.Locked -> EnrollmentProblem.LOCKED
        is MobileReplicationException.Busy -> EnrollmentProblem.BUSY
        is MobileReplicationException.Invalid -> EnrollmentProblem.INVALID
        is MobileReplicationException.StaleRequest -> EnrollmentProblem.CHANGED
        is MobileReplicationException.RecoveryRequired -> EnrollmentProblem.RECOVERY
        is MobileReplicationException.AlreadyConfigured -> EnrollmentProblem.CONFIGURED
        else -> EnrollmentProblem.UNAVAILABLE
    }
    fun cancel(reviewedRequest: ByteArray) {
        if (running) return
        val captured = reviewedRequest.copyOf()
        mutable.value = mutable.value.copy(snapshot = EnrollmentSnapshot(captured, true))
        operate { repository.cancel(captured) }
    }

    private fun operate(action: suspend () -> EnrollmentSnapshot) {
        if (running) return
        running = true
        discardBundleReview()
        mutable.value = mutable.value.copy(loading = true, problem = null)
        viewModelScope.launch {
            try { mutable.value = EnrollmentState(action(), loading = false) }
            catch (cancelled: CancellationException) { throw cancelled }
            catch (error: Exception) {
                mutable.value = mutable.value.copy(loading = false, problem = problemFor(error))
            } finally { running = false }
        }
    }
}

File: apps/ios/TermiRustMobile/Enrollment/EnrollmentViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class EnrollmentViewModel: ObservableObject {
    @Published private(set) var snapshot = EnrollmentSnapshot()
    @Published private(set) var loading = false
    @Published private(set) var problem: LocalizedStringKey?
    private let repository: any EnrollmentRepository

    init(repository: any EnrollmentRepository = NativeEnrollmentRepository()) {
        self.repository = repository
    }

    func reload() async { await operate { try await self.repository.load() } }
    func prepare() async {
        guard problem == nil, snapshot.request == nil else { return }
        await operate { try await self.repository.prepare() }
    }
    func cancel(_ reviewed: Data) async {
        guard !loading else { return }
        snapshot = EnrollmentSnapshot(request: reviewed, cancellationPending: true)
        await operate { try await self.repository.cancel(reviewed) }
    }

    private func operate(_ action: () async throws -> EnrollmentSnapshot) async {
        guard !loading else { return }
        loading = true
        problem = nil
        defer { loading = false }
        do { snapshot = try await action() }
        catch {
            switch error {
            case MobileReplicationError.Locked, ReplicationStorageError.Locked:
                problem = "Unlock your device, then reload or retry cancellation."
            case MobileReplicationError.Busy:
                problem = "Another operation is in progress. Reload to check the saved request."
            case MobileReplicationError.Invalid:
                problem = "Stored enrollment data is invalid. It has not been replaced."
            case MobileReplicationError.StaleRequest:
                problem = "The saved request changed. Reload before continuing."
            case MobileReplicationError.RecoveryRequired:
                problem = "Enrollment recovery is required. Existing data has been preserved."
            case MobileReplicationError.AlreadyConfigured:
                problem = "Replication is already configured. Preparation is unavailable here."
            default:
                problem = "Enrollment storage is unavailable. Reload before trying again."
            }
        }
    }
}
